mention `expect` if we try to `unwrap("string lit")`

### Code

```Rust
fn main() {
    let x = Some(());
    x.unwrap("oh no");
}
```

### Current output

```Shell
error[E0061]: this method takes 0 arguments but 1 argument was supplied
    --> src/main.rs:3:7
     |
   3 |     x.unwrap("oh no");
     |       ^^^^^^ ------- unexpected argument of type `&'static str`
     |
note: method defined here
    --> /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:1013:18
     |
1013 |     pub const fn unwrap(self) -> T {
     |                  ^^^^^^
help: remove the extra argument
     |
   3 -     x.unwrap("oh no");
   3 +     x.unwrap();
     |

For more information about this error, try `rustc --explain E0061`.
```

### Desired output

```Shell
...
help: did you mean to use 'expect'?
     |
   3 -     x.unwrap("oh no");
   3 +     x.expect("oh no");
     |
```

### Rationale and extra context

_No response_

### Other cases

```Rust

```

### Rust Version

```Shell
rustc 1.97.0-nightly (e9e32aca5 2026-04-17)
binary: rustc
commit-hash: e9e32aca5a4ffd08cbc29547b039d64b92a2c03b
commit-date: 2026-04-17
host: x86_64-unknown-linux-gnu
release: 1.97.0-nightly
LLVM version: 22.1.2
```

### Anything else?

_No response_

--- a/compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs
+++ b/compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs
@@ -2342,6 +2342,17 @@ impl<'a, 'b, 'tcx> FnCallDiagCtxt<'a, 'b, 'tcx> {
                         provided_span,
                         format!("unexpected argument{idx}{provided_ty_name}"),
                     ));
+                    if self.provided_arg_tys.len() == 1
+                        && let Some(span) = self.maybe_suggest_expect_for_unwrap(provided_ty)
+                    {
+                        err.span_suggestion_verbose(
+                            span,
+                            "did you mean to use `expect`?",
+                            "expect",
+                            Applicability::MaybeIncorrect,
+                        );
+                        continue;
+                    }
                     let mut span = provided_span;
                     if span.can_be_used_for_suggestions()
                         && self.call_metadata.error_span.can_be_used_for_suggestions()
@@ -2771,6 +2782,22 @@ impl<'a, 'b, 'tcx> FnCallDiagCtxt<'a, 'b, 'tcx> {
         suggestion += ")";
 
         (suggestion_span, suggestion)
+    }
+
+    fn maybe_suggest_expect_for_unwrap(&self, provided_ty: Ty<'tcx>) -> Option<Span> {
+        let tcx = self.tcx();
+        if let Some(call_ident) = self.call_metadata.call_ident
+            && call_ident.name == sym::unwrap
+            && let Some(callee_ty) = self.callee_ty
+            && let ty::Adt(adt, _) = callee_ty.peel_refs().kind()
+            && (tcx.is_diagnostic_item(sym::Option, adt.did())
+                || tcx.is_diagnostic_item(sym::Result, adt.did()))
+            && self.may_coerce(provided_ty, Ty::new_static_str(tcx))
+        {
+            Some(call_ident.span)
+        } else {
+            None
+        }
     }
 }
 

--- a/tests/ui/did_you_mean/expect-instead-of-unwrap.rs
+++ b/tests/ui/did_you_mean/expect-instead-of-unwrap.rs
@@ -1,0 +1,6 @@
+fn main() {
+    Ok(42).unwrap("wow");
+    //~^ ERROR this method takes 0 arguments but 1 argument was supplied
+    Some(42).unwrap("wow");
+    //~^ ERROR this method takes 0 arguments but 1 argument was supplied
+}

--- a/tests/ui/did_you_mean/expect-instead-of-unwrap.stderr
+++ b/tests/ui/did_you_mean/expect-instead-of-unwrap.stderr
@@ -1,0 +1,31 @@
+error[E0061]: this method takes 0 arguments but 1 argument was supplied
+  --> $DIR/expect-instead-of-unwrap.rs:2:12
+   |
+LL |     Ok(42).unwrap("wow");
+   |            ^^^^^^ ----- unexpected argument of type `&'static str`
+   |
+note: method defined here
+  --> $SRC_DIR/core/src/result.rs:LL:COL
+help: did you mean to use `expect`?
+   |
+LL -     Ok(42).unwrap("wow");
+LL +     Ok(42).expect("wow");
+   |
+
+error[E0061]: this method takes 0 arguments but 1 argument was supplied
+  --> $DIR/expect-instead-of-unwrap.rs:4:14
+   |
+LL |     Some(42).unwrap("wow");
+   |              ^^^^^^ ----- unexpected argument of type `&'static str`
+   |
+note: method defined here
+  --> $SRC_DIR/core/src/option.rs:LL:COL
+help: did you mean to use `expect`?
+   |
+LL -     Some(42).unwrap("wow");
+LL +     Some(42).expect("wow");
+   |
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0061`.

